accounts-db/src/file_io.rs

```diff
@@ -0,0 +1,151 @@
+//! File i/o helper functions.
+#[cfg(unix)]
+use std::os::unix::prelude::FileExt;
+use std::{fs::File, ops::Range};
+
+/// `buffer` contains `valid_bytes` of data at its end.
+/// Move those valid bytes to the beginning of `buffer`, then read from `offset` to fill the rest of `buffer`.
+/// Update `offset` for the next read and update `valid_bytes` to specify valid portion of `buffer`.
+/// `valid_len` is # of valid bytes in the file. This may be <= file length.
+pub fn read_more_buffer(
+    file: &File,
+    offset: &mut usize,
+    buffer: &mut [u8],
+    valid_bytes: &mut Range<usize>,
+    valid_len: usize,
+) -> std::io::Result<()> {
+    // copy remainder of `valid_bytes` into beginning of `buffer`.
+    // These bytes were left over from the last read but weren't enough for the next desired contiguous read chunk.
+    buffer.copy_within(valid_bytes.clone(), 0);
+
+    // read the rest of `buffer`
+    let bytes_read = read_buffer(file, *offset, &mut buffer[valid_bytes.len()..], valid_len)?;
+    *offset += bytes_read;
+    *valid_bytes = 0..(valid_bytes.len() + bytes_read);
+
+    Ok(())
+}
+
+#[cfg(unix)]
+/// Read, starting at `start_offset`, until `buffer` is full or we read past `valid_len`/eof.
+/// `valid_len` is # of valid bytes in the file. This may be <= file length.
+/// return # bytes read
+pub fn read_buffer(
+    file: &File,
+    start_offset: usize,
+    buffer: &mut [u8],
+    valid_len: usize,
+) -> std::io::Result<usize> {
+    let mut offset = start_offset;
+    let mut start_read = 0;
+    let mut total_bytes_read = 0;
+    if start_offset >= valid_len {
+        return Ok(0);
+    }
+
+    while start_read < buffer.len() {
+        match file.read_at(&mut buffer[start_read..], offset as u64) {
+            Err(err) => {
+                if err.kind() == std::io::ErrorKind::Interrupted {
+                    continue;
+                }
+                return Err(err);
+            }
+            Ok(bytes_read_this_time) => {
+                total_bytes_read += bytes_read_this_time;
+                if total_bytes_read + start_offset >= valid_len {
```

Comment:
please tell me how to do what you are asking.

Reply:
something like this https://github.com/alessandrod/solana/blob/7a74d5e047433ded5455d5cdf83da274196ed961/accounts-db/src/file_io.rs#L38

Reply:
> something like thi

I mean clamping. I can imagine you'd like the range to be passed in instead of 2 separate fields. I'd prefer to do that after. Perhaps it adds clarity or simplifies. Easy enough to change in isolation without another 10 comments on this pr. ;-)

Reply:
I am clamping in that commit?

Reply:
```
    buffer = &mut buffer[..file_offsets.len().min(buffer_len)];
 ```

This makes sure we never read more than requested, and I updated a test that was checking that we read too much :P

Reply:
oh. I didn't know what you meant by clamping. I was thinking of something like 'pin' since you mentioned performance.

Reply:
I propose what we have is correct and safe (and only enabled by cli arg). I imagine this function and api could be better. I'm awaiting perf results from you. I'd like to make changes to this fn in a different pr. Is that ok?

Reply:
sure